src/github.com/getlantern/detour/detour.go

```diff
@@ -0,0 +1,288 @@
+	readDeadline  time.Time
+	writeDeadline time.Time
+}
+
+func SetTimeout(t time.Duration) {
+	timeoutToDetour = t
+}
+
+func Dialer(dialer dialFunc) dialFunc {
+	return func(network, addr string) (conn net.Conn, err error) {
+		dc := &detourConn{dialDetour: dialer, network: network, addr: addr}
+		if in, _ := inWhitelist(addr); !in {
+			dc.state = stateInitial
+			dc.conn, err = net.DialTimeout(network, addr, timeoutToDetour)
+			if err == nil {
+				log.Tracef("Dialed a new %s connection to %s", dc.stateDesc(), addr)
+				return dc, nil
+			}
+			log.Debugf("Dial %s to %s failed, try detour: %s", dc.stateDesc(), addr, err)
+		}
+		dc.state = stateDetour
+		dc.conn, err = dc.dialDetour(network, addr)
+		if err != nil {
+			log.Errorf("Dial %s to %s failed", dc.stateDesc(), addr)
+			return nil, err
+		}
+		log.Tracef("Dialed a new %s connection to %s", dc.stateDesc(), addr)
+		return dc, err
+	}
+}
+
+// Read() implements the function from net.Conn
+func (dc *detourConn) Read(b []byte) (n int, err error) {
+	conn := dc.getConn()
+	if dc.state != stateInitial {
+		if n, err = conn.Read(b); err != nil {
+			log.Debugf("Read from %s %s failed: %s", dc.addr, dc.stateDesc(), err)
+			if dc.state != stateDetour {
+				log.Debugf("Add %s to white list temporarily", dc.addr)
+				addToWhitelist(dc.addr, false)
+			}
+			return
+		}
+		log.Tracef("Read %d bytes from %s %s", n, dc.addr, dc.stateDesc())
+		return
+	}
+	if in, _ := inWhitelist(dc.addr); in {
+		log.Tracef("%s in white list, detour", dc.addr)
+		return dc.detour(b)
+	}
+	// state will always be settled after first read, safe to clear buffer at end of it
+	defer dc.resetBuffer()
+	dl := time.Now().Add(timeoutToDetour)
+	if !dc.readDeadline.IsZero() && dc.readDeadline.Before(dl) {
+		atomic.CompareAndSwapUint32(&dc.state, stateInitial, stateDirect)
+		n, err = conn.Read(b)
+		log.Tracef("No time left to detour, read %d bytes from %s directly, err=%s", n, dc.addr, err)
+		return
+	}
+
+	log.Tracef("Read from %s directly first", dc.addr)
+	conn.SetReadDeadline(dl)
+	if n, err = conn.Read(b); err != nil {
+		ne, ok := err.(net.Error)
+		if ok && ne.Timeout() {
+			log.Tracef("Timeout read directly from %s, try detour", dc.addr)
+			return dc.detour(b)
+		} else if oe, ok := err.(*net.OpError); ok && oe.Err == syscall.ECONNRESET {
+			log.Tracef("RST received from %s, try detour", dc.addr)
+			return dc.detour(b)
+			/*} else if err == io.EOF && n == 0 {
+			log.Tracef("EOF received with no data from %s, try detour", dc.addr)
+			return dc.detour(b)*/
+		} else {
+			err = fmt.Errorf("Error while read from %s directly: %s", dc.addr, err)
+			log.Error(err)
+			return
+		}
+	}
+	atomic.CompareAndSwapUint32(&dc.state, stateInitial, stateDirect)
+	log.Tracef("Read %d bytes from %s directly, set state to %s", n, dc.addr, dc.stateDesc())
+	return n, nil
+}
+
+// Write() implements the function from net.Conn
+func (dc *detourConn) Write(b []byte) (n int, err error) {
+	if dc.state == stateInitial {
```

Comment:
BTW, a handy way to check if you have possible concurrency issues is to compile/run with the `-race` flag, which enables race detection.